# Project: cleanning - [candy data]
### this notebook is for educational purpose . in this notebook we will do some cleanning

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">wrangling</a></li>
  
</ul>

<a id='intro'></a>
## Introduction

### Dataset Description

> i will go through candy data 2017 , this data has :
- Internal_ID
- Q1-Going Out?: Are you actually going trick or treating yourself? "has values `Yes or No`"
- Q2-Gender: It has four different options, `Femal ,Male ,Other ,I'd rather not say`
- Q3-Age: Numerical field
- Q4-Country: Text Field, but users have written their own version of the names. Example, for America, there are entries such as USA, us, US, America so we should consider that while cleanning
- Q5-State/Province: Text Field, but users have written their own version of the names. Same as the country data.
- Q6-Joy Or Despair: All kinds of chocolate bars are the questions with three distinct options to choose from (Joy, Meh, Despair).
- Q7-Joy Other: Text Field to write items not included above that give you JOY. Lots of missing values.
- Q8-Despair Other: Text Field ti write items not included above that give you DESPAIR. Lots of missing values.
- Q9-Other Comments: Text Field. Lots of missing values.
- Q10-Dress: Binary field. Missing values present
- Unnamed: 113
- Q11-Day: Binary Answer Field. Missing values present


# 1- import packages and load data

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
!pip install fuzzywuzzy

In [3]:
import numpy as np
import pandas as pd
import re
from fuzzywuzzy import process
from fuzzywuzzy import fuzz

/usr/local/lib/python3.10/dist-packages/fuzzywuzzy/fuzz.py:11: UserWarning: Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning
  warnings.warn('Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning')


In [4]:
df = pd.read_excel('/content/drive/MyDrive/ieeetasksfordatascience/Candy Task/candyhierarchy2017.xlsx')

/usr/local/lib/python3.10/dist-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)


<a id='wrangling'></a>
## Data Wrangling



### General Properties

# 2- show some general properties for the data

In [5]:
pd.set_option('display.max_columns', None)

In [6]:
df.head()

Internal ID Q1: GOING OUT? Q2: GENDER Q3: AGE Q4: COUNTRY  \
0     90258773            NaN        NaN     NaN         NaN   
1     90272821             No       Male      44        USA    
2     90272829            NaN       Male      49         USA   
3     90272840             No       Male      40          us   
4     90272841             No       Male      23         usa   

  Q5: STATE, PROVINCE, COUNTY, ETC Q6 | 100 Grand Bar  \
0                              NaN                NaN   
1                               NM                MEH   
2                         Virginia                NaN   
3                               or                MEH   
4                         exton pa                JOY   

  Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                                NaN                                       
1                                            DESPAIR                                       
2                                                NaN                                       
3                                            DESPAIR                                       
4                                            DESPAIR                                       

  Q6 | Any full-sized candy bar Q6 | Black Jacks Q6 | Bonkers (the candy)  \
0                           NaN              NaN                      NaN   
1                           JOY              MEH                  DESPAIR   
2                           NaN              NaN                      NaN   
3                           JOY              MEH                      MEH   
4                           JOY          DESPAIR                      MEH   

  Q6 | Bonkers (the board game) Q6 | Bottle Caps Q6 | Box'o'Raisins  \
0                           NaN              NaN                NaN   
1                       DESPAIR          DESPAIR            DESPAIR   
2                           NaN              NaN                NaN   
3                       DESPAIR              MEH            DESPAIR   
4                       DESPAIR              MEH            DESPAIR   

  Q6 | Broken glow stick Q6 | Butterfinger Q6 | Cadbury Creme Eggs  \
0                    NaN               NaN                     NaN   
1                DESPAIR           DESPAIR                     MEH   
2                    NaN               NaN                     NaN   
3                DESPAIR               MEH                     MEH   
4                DESPAIR               MEH                     MEH   

  Q6 | Candy Corn  \
0             NaN   
1             MEH   
2             NaN   
3         DESPAIR   
4         DESPAIR   

  Q6 | Candy that is clearly just the stuff given out for free at restaurants  \
0                                                NaN                            
1                                            DESPAIR                            
2                                                NaN                            
3                                            DESPAIR                            
4                                            DESPAIR                            

  Q6 | Caramellos Q6 | Cash, or other forms of legal tender Q6 | Chardonnay  \
0             NaN                                       NaN             NaN   
1             MEH                                       JOY             MEH   
2             NaN                                       NaN             NaN   
3             MEH                                       JOY             MEH   
4             JOY                                       MEH             JOY   

  Q6 | Chick-o-Sticks (we don’t know what that is) Q6 | Chiclets  \
0                                              NaN           NaN   
1                                          DESPAIR       DESPAIR   
2                                              NaN           NaN   
3                                              JOY

In [7]:
df.shape

(2460, 120)

In [8]:
df.iloc[:, :40].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2460 entries, 0 to 2459
Data columns (total 40 columns):
 #   Column                                                                                 Non-Null Count  Dtype 
---  ------                                                                                 --------------  ----- 
 0   Internal ID                                                                            2460 non-null   int64 
 1   Q1: GOING OUT?                                                                         2350 non-null   object
 2   Q2: GENDER                                                                             2419 non-null   object
 3   Q3: AGE                                                                                2376 non-null   object
 4   Q4: COUNTRY                                                                            2396 non-null   object
 5   Q5: STATE, PROVINCE, COUNTY, ETC                                                   

In [9]:
df.iloc[:, 40:81].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2460 entries, 0 to 2459
Data columns (total 41 columns):
 #   Column                                                   Non-Null Count  Dtype 
---  ------                                                   --------------  ----- 
 0   Q6 | Hershey’s Milk Chocolate                            1788 non-null   object
 1   Q6 | Hershey's Kisses                                    1782 non-null   object
 2   Q6 | Hugs (actual physical hugs)                         1747 non-null   object
 3   Q6 | Jolly Rancher (bad flavor)                          1766 non-null   object
 4   Q6 | Jolly Ranchers (good flavor)                        1765 non-null   object
 5   Q6 | JoyJoy (Mit Iodine!)                                1434 non-null   object
 6   Q6 | Junior Mints                                        1762 non-null   object
 7   Q6 | Senior Mints                                        1518 non-null   object
 8   Q6 | Kale smoothie                    

In [10]:
df.iloc[:, 81:121].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2460 entries, 0 to 2459
Data columns (total 39 columns):
 #   Column                                                                       Non-Null Count  Dtype  
---  ------                                                                       --------------  -----  
 0   Q6 | Reese’s Peanut Butter Cups                                              1781 non-null   object 
 1   Q6 | Reese's Pieces                                                          1769 non-null   object 
 2   Q6 | Reggie Jackson Bar                                                      1446 non-null   object 
 3   Q6 | Rolos                                                                   1746 non-null   object 
 4   Q6 | Sandwich-sized bags filled with BooBerry Crunch                         1684 non-null   object 
 5   Q6 | Skittles                                                                1754 non-null   object 
 6   Q6 | Smarties (American)                

In [11]:
df.iloc[:, :40].isna().sum()

Internal ID                                                                                  0
Q1: GOING OUT?                                                                             110
Q2: GENDER                                                                                  41
Q3: AGE                                                                                     84
Q4: COUNTRY                                                                                 64
Q5: STATE, PROVINCE, COUNTY, ETC                                                           100
Q6 | 100 Grand Bar                                                                         747
Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)     734
Q6 | Any full-sized candy bar                                                              672
Q6 | Black Jacks                                                                           958
Q6 | Bonkers (the candy)                          

In [12]:
df.iloc[:, 40:81].isna().sum()

Q6 | Hershey’s Milk Chocolate                               672
Q6 | Hershey's Kisses                                       678
Q6 | Hugs (actual physical hugs)                            713
Q6 | Jolly Rancher (bad flavor)                             694
Q6 | Jolly Ranchers (good flavor)                           695
Q6 | JoyJoy (Mit Iodine!)                                  1026
Q6 | Junior Mints                                           698
Q6 | Senior Mints                                           942
Q6 | Kale smoothie                                          744
Q6 | Kinder Happy Hippo                                     945
Q6 | Kit Kat                                                674
Q6 | LaffyTaffy                                             736
Q6 | LemonHeads                                             730
Q6 | Licorice (not black)                                   686
Q6 | Licorice (yes black)                                   685
Q6 | Lindt Truffle                      

In [13]:
df.iloc[:, 81:121].isna().sum()

Q6 | Reese’s Peanut Butter Cups                                                 679
Q6 | Reese's Pieces                                                             691
Q6 | Reggie Jackson Bar                                                        1014
Q6 | Rolos                                                                      714
Q6 | Sandwich-sized bags filled with BooBerry Crunch                            776
Q6 | Skittles                                                                   706
Q6 | Smarties (American)                                                        725
Q6 | Smarties (Commonwealth)                                                    901
Q6 | Snickers                                                                   690
Q6 | Sourpatch Kids (i.e. abominations of nature)                               738
Q6 | Spotted Dick                                                               882
Q6 | Starburst                                                              

In [14]:
df.duplicated().sum()

0

# 3- write summary about what you get form the above functions
> #### tips : this summary can be about : if there is any missing values , if there a wrong data type , if there a wrong range of values , ....
> #### Data has 2460 rows and 120 columns.
> #### Data has zero duplicate.
> #### Q7, Q8, Q9 and "Unnamed: 113" columns have a lot of missing values.
> #### Q6 is the main column in survey.

# cleaning

### 4- drop all the useless columns and  rows with too much NaN ,
> tip : useless columns can be like `Internal ID`and can be also columns with too much NaN
- rename columns for easy access

In [15]:
df.drop(["Internal ID", 'Q7: JOY OTHER', 'Q8: DESPAIR OTHER', 'Q9: OTHER COMMENTS', "Unnamed: 113", "Click Coordinates (x, y)"] , axis="columns", inplace=True)

In [16]:
df.rename(columns = {'Q1: GOING OUT?' : 'Going Out?',
                          'Q2: GENDER' : 'Gender',
                          'Q3: AGE': 'Age',
                          'Q4: COUNTRY' : 'Country',
                          'Q5: STATE, PROVINCE, COUNTY, ETC' : 'Area',
                          'Q10: DRESS' : 'Dress',
                          'Q11: DAY': 'Day',
                          'Q12: MEDIA [Daily Dish]' : 'DailyDish',
                          'Q12: MEDIA [Science]': 'Science',
                          'Q12: MEDIA [ESPN]' : 'ESPN',
                          'Q12: MEDIA [Yahoo]': 'Yahoo'}, inplace=True)

In [17]:
df.columns = [x for x in df.columns.str.replace("Q6 \| ",'')]

<ipython-input-17-0595057fd142>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  df.columns = [x for x in df.columns.str.replace("Q6 \| ",'')]


In [18]:
df.head()

Going Out? Gender  Age Country      Area 100 Grand Bar  \
0        NaN    NaN  NaN     NaN       NaN           NaN   
1         No   Male   44    USA         NM           MEH   
2        NaN   Male   49     USA  Virginia           NaN   
3         No   Male   40      us        or           MEH   
4         No   Male   23     usa  exton pa           JOY   

  Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                                NaN                                  
1                                            DESPAIR                                  
2                                                NaN                                  
3                                            DESPAIR                                  
4                                            DESPAIR                                  

  Any full-sized candy bar Black Jacks Bonkers (the candy)  \
0                      NaN         NaN                 NaN   
1                      JOY         MEH             DESPAIR   
2                      NaN         NaN                 NaN   
3                      JOY         MEH                 MEH   
4                      JOY     DESPAIR                 MEH   

  Bonkers (the board game) Bottle Caps Box'o'Raisins Broken glow stick  \
0                      NaN         NaN           NaN               NaN   
1                  DESPAIR     DESPAIR       DESPAIR           DESPAIR   
2                      NaN         NaN           NaN               NaN   
3                  DESPAIR         MEH       DESPAIR           DESPAIR   
4                  DESPAIR         MEH       DESPAIR           DESPAIR   

  Butterfinger Cadbury Creme Eggs Candy Corn  \
0          NaN                NaN        NaN   
1      DESPAIR                MEH        MEH   
2          NaN                NaN        NaN   
3          MEH                MEH    DESPAIR   
4          MEH                MEH    DESPAIR   

  Candy that is clearly just the stuff given out for free at restaurants  \
0                                                NaN                       
1                                            DESPAIR                       
2                                                NaN                       
3                                            DESPAIR                       
4                                            DESPAIR                       

  Caramellos Cash, or other forms of legal tender Chardonnay  \
0        NaN                                  NaN        NaN   
1        MEH                                  JOY        MEH   
2        NaN                                  NaN        NaN   
3        MEH                                  JOY        MEH   
4        JOY                                  MEH        JOY   

  Chick-o-Sticks (we don’t know what that is) Chiclets Coffee Crisp  \
0                                         NaN      NaN          NaN   
1                                     DESPAIR  DESPAIR      DESPAIR   
2                                         NaN      NaN          NaN   
3                                         JOY      MEH      DESPAIR   
4                                     DESPAIR  DESPAIR          MEH   

  Creepy Religious comics/Chick Tracts Dental paraphenalia Dots Dove Bars  \
0                                  NaN                 NaN  NaN       NaN   
1                              DESPAIR             DESPAIR  MEH       JOY   
2                                  NaN                 NaN  NaN       NaN   
3                              DESPAIR             DESPAIR  JOY       JOY   
4                              DESPAIR             DESPAIR  MEH       JOY   

  Fuzzy Peaches Generic Brand Acetaminophen Glow sticks Goo Goo Clusters  \
0           NaN                         NaN         NaN              NaN   
1       DESPAIR                     DESPAIR     DESPAIR          DESPAIR   
2           NaN                         NaN        

### 5 - clean `going out` column
> tip : you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [19]:
df['Going Out?'].unique()

array([nan, 'No', 'Yes'], dtype=object)

In [20]:
df['Going Out?'].isna().sum()

110

In [21]:
df['Going Out?'] = df['Going Out?'].fillna('Maybe')

In [22]:
df['Going Out?'].value_counts()

No       2038
Yes       312
Maybe     110
Name: Going Out?, dtype: int64

### 6 - clean ` gender` column
> tip : you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [23]:
df['Gender'].unique()

array([nan, 'Male', 'Female', "I'd rather not say", 'Other'], dtype=object)

In [24]:
df['Gender'].isna().sum()

41

In [25]:
df['Gender'] = df['Gender'].fillna("I'd rather not say")

In [26]:
df['Gender'].value_counts()

Male                  1467
Female                 839
I'd rather not say     124
Other                   30
Name: Gender, dtype: int64

### 7 - clean ` age` column
> tip : note the type of age column , and NaN values
 - try to use `to_numeric`
 - tip : you should choose which best value to put in place of these NaNs ( mean, median, ....etc)

In [27]:
df['Age'].unique()

array([nan, 44, 49, 40, 23, 53, 33, 43, 56, 64, 37, 59, 48, 54, 36, 45,
       25, 34, 35, 38, 58, 50, 47, 16, 52, 63, 65, 41, 27, 31, 61, 46, 42,
       62, 29, 39, 32, 28, 69, 67, 30, 22, 26, 51, 70, 24, 18, 19,
       'Old enough', 57, 60, 66, 12, 'Many', 55, 72, '?', 21, 11, 'no', 9,
       68, 20, 6, 10, 71, 90, 13, '45-55', 312, 99, 7, 'hahahahaha', 88,
       39.4, 74, 102, 'older than dirt', 17, 15, 8, 75, '5u', 'Enough',
       'See question 2', '24-50', 14, 'Over 50', 100, 76, 1000,
       'sixty-nine', '46 Halloweens.', 'ancient', 77, 'OLD', 'old', 73,
       70.5, 1, 'MY NAME JEFF', 4, '59 on the day after Halloween',
       'old enough', 'your mom',
       'I can remember when Java was a cool new language', '60+'],
      dtype=object)

In [28]:
df['Age'] = pd.to_numeric(df['Age'] , errors = 'coerce')

In [29]:
df['Age'].isna().sum()

108

In [30]:
df.Age[df['Age']>90]=np.nan

<ipython-input-30-1582d86ec7a9>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.Age[df['Age']>90]=np.nan


In [31]:
df['Age'].mean()

41.9961210571185

In [32]:
df['Age'].median()

41.0

> #### from the mean and the median of the age: Obviously they are almost equal

In [33]:
df['Age'] = df['Age'].fillna(df['Age'].mean())

In [34]:
df['Age'].isna().sum()

0

### 8 - clean ` country` column
> tip : note that country column has country names with different writing ways
  - use fuzzywuzzy to deal with this problem
  - you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [35]:
df['Country'].unique()

array([nan, 'USA ', 'USA', 'us', 'usa', 'canada', 'Canada', 'Us', 'US',
       'Murica', 'United States', 'uk', 'United Kingdom', 'united states',
       'Usa', 'United States ', 'United staes',
       'United States of America', 'UAE', 'England', 'UK', 'canada ',
       'Mexico', 'United states', 'u.s.a.', 'USAUSAUSA', 'america', 35,
       'france', 'United States of America ', 'U.S.A.', 'finland',
       'unhinged states', 'Canada ', 'united states of america',
       'US of A', 'Unites States', 'The United States', 'North Carolina ',
       'Unied States', 'Netherlands', 'germany', 'Europe', 'Earth', 'U S',
       'u.s.', 'U.K. ', 'Costa Rica', 'The United States of America',
       'unite states', 'U.S.', 46, 'cascadia', 'Australia',
       'insanity lately', 'Greece', 'USA? Hard to tell anymore..',
       "'merica", 'usas', 'Pittsburgh', 45, 'United State', 32, 'France',
       'australia', 'A', 'Can', 'Canae', 'New York', 'Trumpistan',
       'Ireland', 'United Sates', 'Korea', 

In [36]:
df['Country'] = df['Country'].str.strip().str.lower()

In [37]:
countries = pd.Series(['usa', 'united states of america','america','united kingdom england', 'hong kong', 'china', 'sweden', 'spain',
                       'ireland', 'taiwan', 'singapore', 'netherlands',
                       'indonesia', 'south korea', 'switzerland', 'denmark', 'scotland', 'canada', 'iceland', 'south africa',
                       'japan', 'greece', 'costa rica', 'germany', 'finland', 'france', 'mexico'])

for country in countries:
    matches = process.extract(country, df['Country'], limit = df.shape[0])
    for potential_match in matches:
        if potential_match[1] >= 80:
            df.loc[df['Country'] == potential_match[0], 'Country'] = country

df['Country'].unique()

array([nan, 'usa', 'canada', 'murica', 'america', 'uk', 'uae',
       'united kingdom england', 'mexico', 'u.s.a.', 'france', 'finland',
       'north carolina', 'netherlands', 'germany', 'europe', 'earth',
       'u.k.', 'costa rica', 'cascadia', 'australia', 'insanity lately',
       'greece', 'usa? hard to tell anymore..', 'pittsburgh', 'canae',
       'new york', 'trumpistan', 'iceland', 'south korea', 'california',
       'japan', 'south africa', 'scotland', 'denmark', 'switzerland',
       'ud', 'new jersey', 'indonesia', 'endland', 'atlantis', 'murrika',
       'alaska', 'soviet canuckistan', 'singapore', 'china', 'taiwan',
       'hong kong', 'spain', 'sweden', 'narnia', 'u s a',
       'subscribe to dm4uz3 on youtube', "i don't know anymore",
       'fear and loathing'], dtype=object)

In [38]:
df['Country'].value_counts()

usa                               1319
america                            726
canada                             226
uk                                  18
germany                             10
u.s.a.                               8
netherlands                          7
australia                            7
scotland                             5
japan                                5
iceland                              5
mexico                               4
switzerland                          3
france                               3
sweden                               2
denmark                              2
south korea                          2
hong kong                            2
china                                2
narnia                               1
spain                                1
u s a                                1
taiwan                               1
ud                                   1
singapore                            1
soviet canuckistan       

In [39]:
mapping = {'murica':'usa', 'united states of america':'usa','america':'usa', 'u.s.a.':'usa',
'north carolina':'usa', 'usa? hard to tell anymore..':'usa', 'pittsburgh':'usa', 'new york':'usa', 'trumpistan':'usa',
          'california':'usa', 'ud':'usa', 'new jersey':'usa', 'endland':'uk', 'soviet canuckistan':'canada',
           'u s a':'usa', 'united kingdom england':'uk', 'u.k.':'uk', 'cascadia':'usa'}
df['Country'] = df['Country'].replace(mapping)

In [40]:
df.Country[~df['Country'].isin(['usa', 'uk','canada'])]='other'

<ipython-input-40-854fb8f5f5a8>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.Country[~df['Country'].isin(['usa', 'uk','canada'])]='other'


In [41]:
df['Country'] = df['Country'].str.upper()
df['Country'].value_counts()

USA       2064
CANADA     227
OTHER      148
UK          21
Name: Country, dtype: int64

### 9 - clean ` area` column
> tip : area column has the same problem as country columns , but looking at all it's unique values may give us another way to deal with it rather than try to fix it

In [42]:
df['Area'].nunique()

511

> #### Area column is not important in the survery due to the extreme difference

In [43]:
df.drop('Area', axis="columns", inplace=True)

In [44]:
df.head()

Going Out?              Gender        Age Country 100 Grand Bar  \
0      Maybe  I'd rather not say  41.996121   OTHER           NaN   
1         No                Male  44.000000     USA           MEH   
2      Maybe                Male  49.000000     USA           NaN   
3         No                Male  40.000000     USA           MEH   
4         No                Male  23.000000     USA           JOY   

  Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                                NaN                                  
1                                            DESPAIR                                  
2                                                NaN                                  
3                                            DESPAIR                                  
4                                            DESPAIR                                  

  Any full-sized candy bar Black Jacks Bonkers (the candy)  \
0                      NaN         NaN                 NaN   
1                      JOY         MEH             DESPAIR   
2                      NaN         NaN                 NaN   
3                      JOY         MEH                 MEH   
4                      JOY     DESPAIR                 MEH   

  Bonkers (the board game) Bottle Caps Box'o'Raisins Broken glow stick  \
0                      NaN         NaN           NaN               NaN   
1                  DESPAIR     DESPAIR       DESPAIR           DESPAIR   
2                      NaN         NaN           NaN               NaN   
3                  DESPAIR         MEH       DESPAIR           DESPAIR   
4                  DESPAIR         MEH       DESPAIR           DESPAIR   

  Butterfinger Cadbury Creme Eggs Candy Corn  \
0          NaN                NaN        NaN   
1      DESPAIR                MEH        MEH   
2          NaN                NaN        NaN   
3          MEH                MEH    DESPAIR   
4          MEH                MEH    DESPAIR   

  Candy that is clearly just the stuff given out for free at restaurants  \
0                                                NaN                       
1                                            DESPAIR                       
2                                                NaN                       
3                                            DESPAIR                       
4                                            DESPAIR                       

  Caramellos Cash, or other forms of legal tender Chardonnay  \
0        NaN                                  NaN        NaN   
1        MEH                                  JOY        MEH   
2        NaN                                  NaN        NaN   
3        MEH                                  JOY        MEH   
4        JOY                                  MEH        JOY   

  Chick-o-Sticks (we don’t know what that is) Chiclets Coffee Crisp  \
0                                         NaN      NaN          NaN   
1                                     DESPAIR  DESPAIR      DESPAIR   
2                                         NaN      NaN          NaN   
3                                         JOY      MEH      DESPAIR   
4                                     DESPAIR  DESPAIR          MEH   

  Creepy Religious comics/Chick Tracts Dental paraphenalia Dots Dove Bars  \
0                                  NaN                 NaN  NaN       NaN   
1                              DESPAIR             DESPAIR  MEH       JOY   
2                                  NaN                 NaN  NaN       NaN   
3                              DESPAIR             DESPAIR  JOY       JOY   
4                              DESPAIR             DESPAIR  MEH       JOY   

  Fuzzy Peaches Generic Brand Acetaminophen Glow sticks Goo Goo Clusters  \
0           NaN                         NaN         NaN              NaN   
1       DESPAIR                     DESPAIR     DESPAIR          DESPAIR 

### 10 - clean ` Q6` columns
>- tip : Q6 has all kinds of chocolate bars and questions with three distinct options to choose from (Joy, Meh, Despair).
- so all the column that has Q6 is a candy name , so you should go through all these column and you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [45]:
df = df.drop(df[df.iloc[:,4:-6].isnull().all(axis=1)].index)

In [46]:
df.shape

(1803, 113)

In [47]:
df['100 Grand Bar'].unique()

array(['MEH', 'JOY', nan, 'DESPAIR'], dtype=object)

In [48]:
df.iloc[:,4:-6]=df.iloc[:,4:-6].fillna('MEH')

In [49]:
df.iloc[:,4:-6].isnull().sum().sum()

0

### 11 - clean ` dress` column
>tip : you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [50]:
df['Dress'].unique()

array(['White and gold', nan, 'Blue and black'], dtype=object)

In [51]:
df['Dress'].value_counts()

White and gold    1079
Blue and black     634
Name: Dress, dtype: int64

In [52]:
df['Dress'] = df['Dress'].fillna('unknown')

In [53]:
df['Dress'].isna().sum()

0

### 12 - clean ` day` column
>tip : you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [54]:
df['Day'].unique()

array(['Sunday', 'Friday', nan], dtype=object)

In [55]:
df['Day'] = df['Day'].fillna('Friday')

In [56]:
df['Day'].isna().sum()

0

In [57]:
df['Day'].value_counts()

Friday    1160
Sunday     643
Name: Day, dtype: int64

> #### I would make the last four columns in a columns named Media

### 11 - Clean ` Media` column

In [58]:
media = df.iloc[:,-4:]
for i in media.columns:
    media.loc[media[i] == 1.0, i] = i
df['Media'] = media[media.columns].apply(lambda x: ''.join(x.dropna().astype(str)),axis=1)
df['Media']=df['Media'].astype('category')
df.drop(['DailyDish', 'Science', 'ESPN', 'Yahoo'] , axis="columns", inplace=True)
df.loc[df['Media'] == '', 'Media'] = df['Media'].mode().iloc[0]
df.head()

Going Out? Gender        Age Country 100 Grand Bar  \
1         No   Male  44.000000     USA           MEH   
3         No   Male  40.000000     USA           MEH   
4         No   Male  23.000000     USA           JOY   
5         No   Male  41.996121   OTHER           JOY   
7         No   Male  33.000000  CANADA           JOY   

  Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
1                                            DESPAIR                                  
3                                            DESPAIR                                  
4                                            DESPAIR                                  
5                                            DESPAIR                                  
7                                            DESPAIR                                  

  Any full-sized candy bar Black Jacks Bonkers (the candy)  \
1                      JOY         MEH             DESPAIR   
3                      JOY         MEH                 MEH   
4                      JOY     DESPAIR                 MEH   
5                      JOY         MEH                 MEH   
7                      JOY     DESPAIR             DESPAIR   

  Bonkers (the board game) Bottle Caps Box'o'Raisins Broken glow stick  \
1                  DESPAIR     DESPAIR       DESPAIR           DESPAIR   
3                  DESPAIR         MEH       DESPAIR           DESPAIR   
4                  DESPAIR         MEH       DESPAIR           DESPAIR   
5                      MEH         MEH           MEH           DESPAIR   
7                      MEH         JOY           MEH               JOY   

  Butterfinger Cadbury Creme Eggs Candy Corn  \
1      DESPAIR                MEH        MEH   
3          MEH                MEH    DESPAIR   
4          MEH                MEH    DESPAIR   
5          JOY            DESPAIR        MEH   
7          JOY                JOY        JOY   

  Candy that is clearly just the stuff given out for free at restaurants  \
1                                            DESPAIR                       
3                                            DESPAIR                       
4                                            DESPAIR                       
5                                                MEH                       
7                                            DESPAIR                       

  Caramellos Cash, or other forms of legal tender Chardonnay  \
1        MEH                                  JOY        MEH   
3        MEH                                  JOY        MEH   
4        JOY                                  MEH        JOY   
5        JOY                                  JOY        JOY   
7        JOY                                  JOY        JOY   

  Chick-o-Sticks (we don’t know what that is) Chiclets Coffee Crisp  \
1                                     DESPAIR  DESPAIR      DESPAIR   
3                                         JOY      MEH      DESPAIR   
4                                     DESPAIR  DESPAIR          MEH   
5                                         MEH      MEH          JOY   
7                                     DESPAIR      MEH          MEH   

  Creepy Religious comics/Chick Tracts Dental paraphenalia Dots Dove Bars  \
1                              DESPAIR             DESPAIR  MEH       JOY   
3                              DESPAIR             DESPAIR  JOY       JOY   
4                              DESPAIR             DESPAIR  MEH       JOY   
5                              DESPAIR                 MEH  MEH       JOY   
7                              DESPAIR             DESPAIR  MEH       MEH   

  Fuzzy Peaches Generic Brand Acetaminophen Glow sticks Goo Goo Clusters  \
1       DESPAIR                     DESPAIR     DESPAIR          DESPAIR   
3           MEH                         MEH         JOY              JOY   
4       DESPAIR                     DESPAIR     DESPAIR              MEH   

### 13 - check
> after finish cleaning you should make sure your data is clean and that :
- there is no wrong data type
- there is no NaN values
- all the column are cleaing and ready for the next step `EDA`

In [59]:
df.isna().sum().sum()

0

In [60]:
df['Age'] = df['Age'].astype(int)

In [61]:
df.head(1)

Going Out? Gender  Age Country 100 Grand Bar  \
1         No   Male   44     USA           MEH   

  Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
1                                            DESPAIR                                  

  Any full-sized candy bar Black Jacks Bonkers (the candy)  \
1                      JOY         MEH             DESPAIR   

  Bonkers (the board game) Bottle Caps Box'o'Raisins Broken glow stick  \
1                  DESPAIR     DESPAIR       DESPAIR           DESPAIR   

  Butterfinger Cadbury Creme Eggs Candy Corn  \
1      DESPAIR                MEH        MEH   

  Candy that is clearly just the stuff given out for free at restaurants  \
1                                            DESPAIR                       

  Caramellos Cash, or other forms of legal tender Chardonnay  \
1        MEH                                  JOY        MEH   

  Chick-o-Sticks (we don’t know what that is) Chiclets Coffee Crisp  \
1                                     DESPAIR  DESPAIR      DESPAIR   

  Creepy Religious comics/Chick Tracts Dental paraphenalia Dots Dove Bars  \
1                              DESPAIR             DESPAIR  MEH       JOY   

  Fuzzy Peaches Generic Brand Acetaminophen Glow sticks Goo Goo Clusters  \
1       DESPAIR                     DESPAIR     DESPAIR          DESPAIR   

  Good N' Plenty Gum from baseball cards Gummy Bears straight up Hard Candy  \
1            MEH                 DESPAIR                     MEH        MEH   

  Healthy Fruit Heath Bar Hershey's Dark Chocolate Hershey’s Milk Chocolate  \
1       DESPAIR       MEH                      JOY                      JOY   

  Hershey's Kisses Hugs (actual physical hugs) Jolly Rancher (bad flavor)  \
1              MEH                     DESPAIR                    DESPAIR   

  Jolly Ranchers (good flavor) JoyJoy (Mit Iodine!) Junior Mints Senior Mints  \
1                          MEH              DESPAIR      DESPAIR      DESPAIR   

  Kale smoothie Kinder Happy Hippo Kit Kat LaffyTaffy LemonHeads  \
1       DESPAIR            DESPAIR     JOY    DESPAIR        MEH   

  Licorice (not black) Licorice (yes black) Lindt Truffle Lollipops     Mars  \
1                  MEH                  JOY           MEH   DESPAIR  DESPAIR   

  Maynards Mike and Ike Milk Duds Milky Way Regular M&Ms Peanut M&M’s  \
1  DESPAIR          MEH       MEH       JOY          JOY          MEH   

  Blue M&M's Red M&M's Green Party M&M's Independent M&M's  \
1        JOY       JOY               JOY               JOY   

  Abstained from M&M'ing. Minibags of chips Mint Kisses Mint Juleps  \
1                 DESPAIR           DESPAIR         MEH     DESPAIR   

  Mr. Goodbar Necco Wafers    Nerds Nestle Crunch Now'n'Laters    Peeps  \
1     DESPAIR      DESPAIR  DESPAIR           JOY      DESPAIR  DESPAIR   

   Pencils Pixy Stix Real Housewives of Orange County Season 9 Blue-Ray  \
1  DESPAIR   DESPAIR                                            DESPAIR   

  Reese’s Peanut Butter Cups Reese's Pieces Reggie Jackson Bar Rolos  \
1                        JOY            JOY            DESPAIR   JOY   

  Sandwich-sized bags filled with BooBerry Crunch Skittles  \
1                                         DESPAIR  DESPAIR   

  Smarties (American) Smarties (Commonwealth) Snickers  \
1             DESPAIR                 DESPAIR      MEH   

  Sourpatch Kids (i.e. abominations of nature) Spotted Dick Starburst  \
1                                      DESPAIR      DESPAIR       MEH   

  Sweet Tarts Swedish Fish Sweetums (a friend to diabetes)   Take 5 Tic Tacs  \
1     DESPAIR          MEH                         DESPAIR  DESPAIR  DESPAIR   

  Those odd marshmallow circus peanut things Three Musketeers  \
1                                    DESPAIR              JOY   

  Tolberone something or other Trail Mix Twix  \
1                          JOY   DESPAIR  JOY   

  Vials of pure high fructose corn syrup, for 

In [62]:
df.shape

(1803, 110)

### 14 - write a summary about all the cleaning steps
> #### clean each column from invalid data.
> #### fill the NaN values in each column with the suitable value.
> #### drop some columns and rows that was not important and the survery are not affected by them.
> #### overall there are `110` rows and `1803` columns.
> #### there are about `103` columns which are the main candy columns.
> #### the last `4` columns I mixed them into a column called `Media`.